### ISSUE TITLE
istio/tools/license needs to be updated for Go modules

### ISSUE THREAD

**Issue by geeknoid** [istio/istio#14475]
The license checker needs to be updated to work for Go modules. Once that's done, the license checker then needs to be deployed to all Go branches that contribute binaries to our distribution. For now, this would be istio/operator and istio/istio.

**Comment by ericvn** [comment]
Getting back on some of the issues related to using modules during the build. 

Looking at the current license code, it's basically a `go list -f '{{ join .Deps  "\n"}}' ./vendor/...` to generate a long list of packages. The list is then deduped and all entries not starting with `Istio.io/istio/vendor` are removed (some other minor stuff but not listed for simplicity). This yields basically a list of packages in the vendor directory. 

Each package is checked for a license file (we move up the path to find a repos license file) and some magic is done using `licensee` against the license files. A list of those packages without license files is also, kept. etc.

Now, how to handle this when using go modules? Without a vendor directory?

First thought is to simply generate the `vendor` directory using `go mod vendor` and use that with the existing code. That doesn't seem like a really good solution.

Thought is instead of listing the packages, to instead list the modules, removing `istio.io/istio`. Then iterate through the modules finding all the license files (set max recurse to 8?) and then process each license file generating a similar output to the original.

Possible problems:
 - A module hasn't been pulled into the cache so we can't check it for license files
   - How do we make sure that `go build/test` had reason to download a module into the cache. (example missing on a local machine:
```/Users/ericvn/work/go/pkg/mod/github.com/!azure/go-ansiterm@v0.0.0-20170929234023-d6e3b3328b78
/Users/ericvn/work/go/pkg/mod/github.com/!azure/go-autorest@v11.1.0+incompatible
/Users/ericvn/work/go/pkg/mod/github.com/!burnt!sushi/toml@v0.3.1
/Users/ericvn/work/go/pkg/mod/github.com/!data!dog/datadog-go@v2.2.0+incompatible
/Users/ericvn/work/go/pkg/mod/github.com/!jeffail/gabs@v1.1.1
/Users/ericvn/work/go/pkg/mod/github.com/!masterminds/semver@v1.4.0
/Users/ericvn/work/go/pkg/mod/github.com/!masterminds/sprig@v2.14.1+incompatible
/Users/ericvn/work/go/pkg/mod/github.com/!microsoft/go-winio@v0.4.12
/Users/ericvn/work/go/pkg/mod/github.com/!nvveen/!gotty@v0.0.0-20120604004816-cd527374f1e5
/Users/ericvn/work/go/pkg/mod/github.com/!s!a!p/go-hdb@v0.14.1
/Users/ericvn/work/go/pkg/mod/github.com/!sermo!digital/jose@v0.9.1
```
- Are we missing a license file within the cached directory?
  - Since we aren't listing all packages, we don't know that a particular package doesn't have a license file. We could employ some magic to say the there must be a license file within every module. This is similar to the old code where we searched up from the package dir to find what would be a module (or sub-module license).

**Comment by geeknoid** [comment]
"go mod download" forces download into the cache.

Given that we're importing modules, I don't think we need to be concerned about packages. Can we just iterate through all modules and try to find a LICENSE file at the top of each module's cache directory?

**Comment by ericvn** [comment]
@geeknoid I'm fine with simplifying this to be checking for a single LICENSE file at the top of the module (and not searching for additional LICENSE files lower in the module). I can actually code that up and see what it reports for those missing LICENSE files.

**Comment by ericvn** [comment]
One thing I noticed when updating code to try and use the $GOPATH/pkg/mod list that it contains more files than what vendor contains. This means that the _find_ will find additional files in the module case (like license_test.go) which will affect the results (as we only look at the first license file today). We could make the _find_ smarter and also add support for multiple license files in a package. 

As an aside, @sdake was looking to see if there is already some upstream Google tooling that has been written.

**Comment by ericvn** [comment]
There is discussion on what is/isn't copied in the go mode vendor in https://github.com/golang/go/issues/26366 and some related issues. It's nice that the vendor is smaller than the module in the case where it's checked in. Just not consistent since you now have more LICENSE files found in the module case.

**Comment by geeknoid** [comment]
How's this going?

I was just looking at our existing license checking tool and I don't understand all the complicated stuff it is doing around running git commands and such. I believe this tool can be simplified considerably.

Given a repo, you walk up to it and run the tool. The tool reads the go.mod file and for each module, finds the directory in the Go module cache. At the root of the directory, read the LICENSE file. See if it is a supported license (same check the existing tool does today), and if so, move on.

Am I missing some complication here?

**Comment by ostromart** [comment]
The git stuff is there because an external caller may want to generate licenses where version of the tool you need > version of the repo with the licenses and it happens that the tool is in the same repo. The solution to that could be to move the tool out to a separate repo (a good idea anyway IMO). 
There are a few other annoyances with licenses which this tool takes care of. One is that we often import a package some ways into a module, and the package dir doesn't contain the license so you have to walk up the tree. That can be removed if using module granularity. 
In short, git and license searching parts can probably be removed. The main things to keep are the mapping to Google approved licenses, the whitelist for manually verified packages and the various tool modes which serve different purposes (generating license bundle for release, presubmit check, license list for legal review).

**Comment by ericvn** [comment]
@geeknoid, It is more complicated than that because the go module cache contains a lot more files than the _vendor_ directory does. _go mod vendor_ similar to _dep_ creates a smaller _vendor_ directory that doesn't include unused packages within a repo and also removed test files as well as some other files.

I could easily hardcore it to look for a `LICENSE` (instead of license* ignoring case with will find a file like license_test.go which really isn't a license file) or COPYING and use one file only. The negative of doing this would be a case where a package had another LICENSE file lower in the directory structure that we would miss. Is that an issue?

**Comment by geeknoid** [comment]
I propose just looking at files called LICENSE at the root of all dependent
modules, and not try to find files by any other name or path.  Start with
that and if any modules don't have LICENSE file, then we flag those and
figure out what the right strategy is. I bet they'll all have a LICENSE
file at the top.

**Comment by ericvn** [comment]
@geeknoid. I will code that up shortly and generate a list of those modules which have no LICENSE or COPYING and will use the top level file to call `licensee`. Will see what that provides.

**Comment by ericvn** [comment]
Code changes made and a WIP PR coming soon. @geeknoid, is there a preference where this code should end? I suspect we want it moved to _istio/tools_.  Do we may want to move in subsequent PRs if someone wants to actually see the diff between the package based checker and the module based checker. My thought is a PR against _istio/istio_ with the change so it can be reviewed and then subsequent PRs to actually move it if needed.

Some data comparison:

--check before yields:
```
===========================================================
The following packages have reciprocal licenses (code may
be used but not modified):
===========================================================
github.com/hashicorp/hcl/hcl/ast
github.com/hashicorp/go-multierror
github.com/hashicorp/hcl/json/token
github.com/hashicorp/golang-lru/simplelru
github.com/hashicorp/hcl/json/parser
github.com/hashicorp/serf/coordinate
github.com/hashicorp/vault/helper/compressutil
github.com/hashicorp/vault/helper/parseutil
github.com/hashicorp/hcl/json/scanner
github.com/hashicorp/go-rootcerts
github.com/hashicorp/hcl/hcl/token
github.com/hashicorp/hcl/hcl/scanner
github.com/hashicorp/hcl/hcl/printer
github.com/hashicorp/hcl/hcl/parser
github.com/hashicorp/golang-lru
github.com/hashicorp/vault/helper/strutil
github.com/hashicorp/vault/helper/jsonutil
github.com/hashicorp/go-cleanhttp
github.com/hashicorp/go-retryablehttp
github.com/hashicorp/hcl/hcl/strconv
github.com/hashicorp/hcl
github.com/hashicorp/errwrap
===========================================================
The following packages have unknown status licenses (legal
review required). 
===========================================================
github.com/gophercloud/gophercloud/pagination:ECL-2.0
github.com/ghodss/yaml:UNKNOWN
github.com/antlr/antlr4/runtime/Go/antlr:UNKNOWN
github.com/Masterminds/sprig:UNKNOWN
github.com/gophercloud/gophercloud/openstack/utils:ECL-2.0
sigs.k8s.io/yaml:UNKNOWN
github.com/gophercloud/gophercloud:ECL-2.0
github.com/gophercloud/gophercloud/openstack/identity/v2/tokens:ECL-2.0
github.com/gophercloud/gophercloud/openstack:ECL-2.0
github.com/gophercloud/gophercloud/openstack/identity/v3/tokens:ECL-2.0
github.com/gophercloud/gophercloud/openstack/identity/v2/tenants:ECL-2.0
github.com/magiconair/properties:UNKNOWN
github.com/russross/blackfriday:UNKNOWN
github.com/pmezard/go-difflib/difflib:UNKNOWN
github.com/Masterminds/semver:UNKNOWN
```
while after yields:
```
===========================================================
The following packages have reciprocal licenses (code may
be used but not modified):
===========================================================
github.com/armon/consul-api
github.com/go-sql-driver/mysql
github.com/hashicorp/consul
github.com/hashicorp/errwrap
github.com/hashicorp/go-cleanhttp
github.com/hashicorp/go-immutable-radix
github.com/hashicorp/go-memdb
github.com/hashicorp/go-multierror
github.com/hashicorp/go-plugin
github.com/hashicorp/go-retryablehttp
github.com/hashicorp/go-rootcerts
github.com/hashicorp/go-sockaddr
github.com/hashicorp/go-uuid
github.com/hashicorp/go-version
github.com/hashicorp/golang-lru
github.com/hashicorp/hcl
github.com/hashicorp/memberlist
github.com/hashicorp/serf
github.com/hashicorp/vault
github.com/hashicorp/yamux
github.com/jefferai/jsonx
===========================================================
The following packages have unknown status licenses (legal
review required). 
===========================================================
github.com/Masterminds/semver:UNKNOWN
github.com/Masterminds/sprig:UNKNOWN
github.com/alicebob/gopher-json:UNKNOWN
github.com/antlr/antlr4:UNKNOWN
github.com/duosecurity/duo_api_golang:UNKNOWN
github.com/ghodss/yaml:UNKNOWN
github.com/gogo/protobuf:UNKNOWN
github.com/google/go-github:UNKNOWN
github.com/gophercloud/gophercloud:ECL-2.0
github.com/gotestyourself/gotestyourself:UNKNOWN
github.com/inconshreveable/mousetrap:UNKNOWN
github.com/jmespath/go-jmespath:UNKNOWN
github.com/julienschmidt/httprouter:UNKNOWN
github.com/magiconair/properties:UNKNOWN
github.com/miekg/dns:UNKNOWN
github.com/pascaldekloe/goe:UNKNOWN
github.com/pmezard/go-difflib:UNKNOWN
github.com/russross/blackfriday:UNKNOWN
gopkg.in/check.v1:UNKNOWN
gopkg.in/mgo.v2:UNKNOWN
gopkg.in/tomb.v1:BSD-3-Clause-Clear
gotest.tools:UNKNOWN
sigs.k8s.io/yaml:UNKNOWN
```

Let me discuss some differences:
  - In the before there would be multiple lines for a given module because the module has license files in lower directories. ex: 
    ```
    github.com/hashicorp/hcl/hcl/token
    github.com/hashicorp/hcl/hcl/scanner
    github.com/hashicorp/hcl/hcl/printer
    github.com/hashicorp/hcl/hcl/parser
    ```
vs after where we only have `github.com/hashicorp/hcl`
  - In the after case there was license listed for packages that didn't exist in the before since the dependency tooling know to not move those paths to _vendor_. ex:
    ```
    github.com/hashicorp/errwrap
    github.com/hashicorp/go-cleanhttp
    github.com/hashicorp/go-immutable-radix
    github.com/hashicorp/go-memdb
    ```
  - One other bit of weirdness I haven't exactly figured out is that we have a modules like _github.com/armon/consul-api_ which exist in our go.sum and shows up in the after output but isn't in the _vedor_ directory or before input.

**Comment by ericvn** [comment]
Another weirdness in the **after** is a package like _github.com/alicebob/gopher-json_. In the **before** I see it's checked but doesn't return an unknown type while in the **after** it must.

**Comment by geeknoid** [comment]
Yes, the tool should eventually land in istio/tools and be published in a container we can use from any repo. But starting in istio/istio seems fine.

**Comment by ericvn** [comment]
One other piece of info I'm not sure how to handle. The default output looks something like this:
```
-----------------------------------------------------------
Package: sigs.k8s.io/yaml
License URL: /Users/ericvn/work/go/pkg/mod/sigs.k8s.io/yaml@v1.1.0/LICENSE
License text:
The MIT License (MIT)

Copyright (c) 2014 Sam Ghods

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
.....
```
where the License URL field in the **before** case pointed to something like
```
License URL: https://github.com/istio/istio/blob/6e0a06a0b5a7393acde4657cdb9b0d14b6a78a0b/vendor/sigs.k8s.io/yaml/LICENSE
```

So in the **before** case we actually have a link that would also show the contents of the file (it is included above so maybe it's not that important), while there is no real way to create a valid URL in the **after** case since AFAIK there is no way to get the specific URL for the module. In the above case, you could go to _https://github.com/kubernetes-sigs/yaml/releases_ grab the commit for v1.1.0 and then look at the LICENSE file.  Again, the actual contents are included so maybe it's not so important.

**Comment by ericvn** [comment]
What I think I'll do (assuming no one needs the specific URL label is to change the label to Module path and give the path of the file relative to the module cache root, so `mod/sigs.k8s.io/yaml@v1.1.0/LICENSE` in the above example.

**Comment by ericvn** [comment]
#15595 is ready for review.

**Comment by ericvn** [comment]
Closing issue with the merging of #15595

**Comment by ericvn** [comment]
I believe this can be closed with the merging of #15595
